src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala

```diff
@@ -116,10 +116,16 @@ object TCPPeerGroup {
     override val to: InetMultiAddress = InetMultiAddress(nettyChannel.remoteAddress())
 
     override def sendMessage(message: M): Task[Unit] = {
-      toTask({
-        nettyChannel
-          .writeAndFlush(Unpooled.wrappedBuffer(codec.encode(message)))
-      })
+      val byteBuf = Unpooled.wrappedBuffer(codec.encode(message))
+      try {
+        toTask({
+          nettyChannel
+            .writeAndFlush(byteBuf)
+        })
+      } finally {
+        byteBuf.release()
```

Comment:
or even better, using special function for resources management:
```
Task(Unpooled.wrappedBuffer(codec.encode(message))) { byteBuf =>
  nettyChannel.writeAndFlush(byteBuf)
}(byteBuf => Task(byteBuf.release()))
```

Reply:
In fact, writeAndFlush releases the buffer. The writeAndFlush call returns a ChannelFuture and executes asynchronously. When wrapped in toTask so that it has completed, you receive an IllegalReferenceCountException from netty (because our release call tries to decrement the ref count to -1). I have updated these writeAndFlush spots accordingly.